File: src/0783.minimum-distance-between-bst-nodes.799/solution.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor (val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.left = (left === undefined ? null : left);
        this.right = (right === undefined ? null : right);
    }
}

function minDiffInBST (root: TreeNode | null): number {
    if (root === null) {
        return 0;
    }
    const seq:number[] = [];
    inorder(root, seq);
    let result:number = seq[1] - seq[0];
    for (let i = 2; i < seq.length; i++) {
        result = Math.min(seq[i] - seq[i - 1], result);
    }

    return result;
}

function inorder (root:TreeNode, result:number[]):void {
    root.left !== null && inorder(root.left, result);
    result.push(root.val);
    root.right !== null && inorder(root.right, result);
}

File: src/0783.minimum-distance-between-bst-nodes.799/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int last = Integer.MAX_VALUE;
    private int result = Integer.MAX_VALUE;

    public int minDiffInBST(TreeNode root) {
        last = root.val;
        inOrder(root);
        return result;
    }

    private void inOrder(TreeNode node) {
        if (node.left != null) {
            inOrder(node.left);
        }

        if (last < node.val) {
            result = Math.min(result, node.val - last);
        }

        last = node.val;

        if (node.right != null) {
            inOrder(node.right);
        }

    }
}